File: clients/go/create_partitions_test.go
package volant_test

import (
	"errors"
	"net"
	"testing"
	"time"

	volant "github.com/volant-mq/volant/clients/go"
	"github.com/volant-mq/volant/clients/go/codec"
	"github.com/volant-mq/volant/clients/go/frame"
)

type createPartitionsServerResult struct {
	topic      string
	totalCount uint32
	err        error
}

func serveCreatePartitions(t *testing.T, errorCode uint16, partitions uint32) (addr string, got *createPartitionsServerResult, stop func()) {
	t.Helper()
	ln, err := net.Listen("tcp", "127.0.0.1:0")
	if err != nil {
		t.Fatal(err)
	}
	res := &createPartitionsServerResult{}
	done := make(chan struct{})
	go func() {
		defer close(done)
		conn, err := ln.Accept()
		if err != nil {
			res.err = err
			return
		}
		defer conn.Close()
		_ = conn.SetDeadline(time.Now().Add(5 * time.Second))
		buf := make([]byte, 0, 4096)
		tmp := make([]byte, 4096)
		for {
			f, rest, err := frame.TryDecode(buf)
			if err != nil {
				res.err = err
				return
			}
			if f == nil {
				n, err := conn.Read(tmp)
				if n > 0 {
					buf = append(buf, tmp[:n]...)
				}
				if err != nil {
					res.err = err
					return
				}
				continue
			}
			buf = append([]byte(nil), rest...)
			if f.Opcode != codec.OpCreatePartitions {
				res.err = &frame.ProtocolError{Msg: "unexpected opcode"}
				return
			}
			req, err := codec.DecodeCreatePartitionsRequest(f.Payload)
			if err != nil {
				res.err = err
				return
			}
			res.topic = req.Topic
			res.totalCount = req.TotalCount
			newTotal := partitions
			if errorCode != 0 {
				newTotal = 0
			}
			payload, err := codec.EncodeCreatePartitionsResponse(codec.CreatePartitionsResponse{
				ErrorCode:  errorCode,
				Topic:      req.Topic,
				Partitions: newTotal,
			})
			if err != nil {
				res.err = err
				return
			}
			raw, err := frame.Encode(codec.OpCreatePartitionsResponse, f.CorrelationID, payload)
			if err != nil {
				res.err = err
				return
			}
			if _, err := conn.Write(raw); err != nil {
				res.err = err
				return
			}
		}
	}()
	return ln.Addr().String(), res, func() {
		_ = ln.Close()
		select {
		case <-done:
		case <-time.After(2 * time.Second):
		}
	}
}

func TestCreatePartitionsSuccessReturnsNewCount(t *testing.T) {
	addr, got, stop := serveCreatePartitions(t, 0, 4)
	defer stop()
	c, err := volant.DialTimeout(addr, 5*time.Second)
	if err != nil {
		t.Fatal(err)
	}
	defer c.Close()
	out, err := c.CreatePartitions("events", 4)
	if err != nil {
		t.Fatal(err)
	}
	if got.err != nil {
		t.Fatal(got.err)
	}
	if got.topic != "events" || got.totalCount != 4 {
		t.Fatalf("wire request topic=%q totalCount=%d", got.topic, got.totalCount)
	}
	if out != 4 {
		t.Fatalf("parsed %d", out)
	}
}

func TestCreatePartitionsNonzeroErrorRaises(t *testing.T) {
	addr, _, stop := serveCreatePartitions(t, 2, 0)
	defer stop()
	c, err := volant.DialTimeout(addr, 5*time.Second)
	if err != nil {
		t.Fatal(err)
	}
	defer c.Close()
	_, err = c.CreatePartitions("missing", 4)
	if err == nil {
		t.Fatal("expected error")
	}
	var be *volant.BrokerError
	if !errors.As(err, &be) {
		t.Fatalf("got %T %v", err, err)
	}
	if be.Code != 2 || be.Op != "create_partitions" {
		t.Fatalf("broker error %+v", be)
	}
}

File: clients/java/src/test/java/io/volant/CreatePartitionsTest.java
package io.volant;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;

import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.net.InetAddress;
import java.net.ServerSocket;
import java.net.Socket;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.concurrent.atomic.AtomicReference;
import org.junit.jupiter.api.Test;

/** CreatePartitions client tests against a scripted TCP broker (no live server). */
class CreatePartitionsTest {
    @Test
    void successReturnsNewCount() throws Exception {
        try (CreatePartitionsServer srv = CreatePartitionsServer.ok(4)) {
            try (Client c = Client.connect("127.0.0.1", srv.port, 5_000)) {
                int got = c.createPartitions("events", 4);
                assertEquals(4, got);
            }
            assertEquals("events", srv.topic.get());
            assertEquals(4, srv.totalCount.get());
        }
    }

    @Test
    void nonzeroErrorCodeRaises() throws Exception {
        try (CreatePartitionsServer srv = CreatePartitionsServer.error(2)) {
            try (Client c = Client.connect("127.0.0.1", srv.port, 5_000)) {
                BrokerException ex =
                        assertThrows(BrokerException.class, () -> c.createPartitions("missing", 4));
                assertEquals(2, ex.code);
                assertEquals("create_partitions", ex.op);
            }
        }
    }

    private static final class CreatePartitionsServer implements AutoCloseable {
        final int port;
        final AtomicReference<String> topic = new AtomicReference<>();
        final AtomicInteger totalCount = new AtomicInteger();
        private final int errorCode;
        private final int partitions;
        private final ServerSocket listen;
        private final Thread thread;
        private final AtomicReference<Exception> error = new AtomicReference<>();

        static CreatePartitionsServer ok(int partitions) throws IOException {
            return new CreatePartitionsServer(0, partitions);
        }

        static CreatePartitionsServer error(int code) throws IOException {
            return new CreatePartitionsServer(code, 0);
        }

        private CreatePartitionsServer(int errorCode, int partitions) throws IOException {
            this.errorCode = errorCode;
            this.partitions = partitions;
            listen = new ServerSocket(0, 8, InetAddress.getByName("127.0.0.1"));
            listen.setSoTimeout(8_000);
            port = listen.getLocalPort();
            thread = new Thread(this::serve, "volant-create-partitions");
            thread.setDaemon(true);
            thread.start();
        }

        private void serve() {
            try (Socket conn = listen.accept()) {
                conn.setSoTimeout(5_000);
                InputStream in = conn.getInputStream();
                OutputStream out = conn.getOutputStream();
                byte[] buf = new byte[0];
                while (true) {
                    Frame.Decode d = Frame.tryDecode(buf);
                    if (d.frame == null) {
                        byte[] tmp = new byte[4096];
                        int n = in.read(tmp);
                        if (n < 0) {
                            return;
                        }
                        byte[] next = new byte[buf.length + n];
                        System.arraycopy(buf, 0, next, 0, buf.length);
                        System.arraycopy(tmp, 0, next, buf.length, n);
                        buf = next;
                        continue;
                    }
                    buf = d.rest;
                    if (d.frame.opcode != Codec.OP_CREATE_PARTITIONS) {
                        error.set(new ProtocolException("unexpected opcode " + d.frame.opcode));
                        return;
                    }
                    Codec.CreatePartitionsRequest req =
                            Codec.decodeCreatePartitionsRequest(d.frame.payload);
                    topic.set(req.topic);
                    totalCount.set((int) req.totalCount);
                    int newTotal = errorCode != 0 ? 0 : partitions;
                    byte[] payload = Codec.encodeCreatePartitionsResponse(
                            new Codec.CreatePartitionsResponse(errorCode, req.topic, newTotal));
                    out.write(
                            Frame.encode(
                                    Codec.OP_CREATE_PARTITIONS_RESPONSE,
                                    d.frame.correlationId,
                                    payload));
                    out.flush();
                }
            } catch (Exception e) {
                error.set(e);
            }
        }

        @Override
        public void close() {
            try {
                listen.close();
            } catch (IOException ignored) {
                // best-effort
            }
            try {
                thread.join(2_000);
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
            }
        }
    }
}
